base/inference.jl

```diff
@@ -2120,13 +2121,139 @@ function inlineable(f, e::Expr, atypes, sv, enclosing_ast)
     if isa(f,IntrinsicFunction)
         return NF
     end
+    if is(f, invoke)
+        return inlineable_invoke(f, e, atypes, sv, enclosing_ast, argexprs)
+    end
+
+    return inlineable_gf(f, e, atypes, sv, enclosing_ast, argexprs)
+end
+
+function get_invoke_types{T<:Top}(ts::Type{Type{T}})
+    return T
+end
+
+function get_invoke_types(ts::Tuple)
+    return tuple(Any[get_invoke_types(t) for t in ts]...)::Type
+end
+
+function inlineable_invoke(f::Function, e::Expr, atypes, sv,
+                           enclosing_ast, argexprs)
+    f = isconstantfunc(argexprs[1], sv)
+    if is(f, false)
+        return NF
+    end
+    f = _ieval(f)
+    local invoke_types
+    try
+        invoke_types = get_invoke_types(atypes[2])
+    catch err
+        return NF
+    end
 
+    argexprs = argexprs[3:end]
+    atypes = atypes[3:end]
+    if length(atypes) != length(invoke_types)
+        return NF
+    end
+    if isleaftype(invoke_types) && invoke_types == atypes
+        new_e = Expr(:call, f, argexprs...)
+        new_e.typ = e.typ
+        res = inlineable_gf(f, new_e, invoke_types, sv,
+                            enclosing_ast, argexprs)
+        return is(res, NF) ? (new_e, []) : res
+    end
+
+    local meth
+    try
+        meth = ccall(:jl_gf_invoke_lookup, Any, (Any, Any), f, invoke_types)
+    catch
+        return NF
+    end
+    if is(meth, nothing)
+        return NF
+    end
+
+    stmts = []
+    check_stmts = []
+    atypes_l = Type[atypes...]
+    err_label = genlabel(sv)
+    after_err_label = genlabel(sv)
+    for i in 1:length(atypes_l)
+        arg_name = unique_name(enclosing_ast)
+        tmp_ex = :($arg_name = $(argexprs[i]))
+        tmp_ex.typ = atypes_l[i]
+        add_variable(enclosing_ast, arg_name, atypes_l[i], true)
+        push!(stmts, tmp_ex)
+        if !issubtype(atypes_l[i], invoke_types[i])
+            atypes_l[i] = typeintersect(atypes_l[i], invoke_types[i])
+            check_type = :($isa($arg_name, $(invoke_types[i])))
+            check_type.typ = Bool
+            push!(check_stmts, Expr(:gotoifnot, check_type, err_label.label))
+        end
+        argexprs[i] = SymbolNode(arg_name, atypes_l[i])
+    end
+    if !isempty(check_stmts)
+        append!(stmts, check_stmts)
+        push!(stmts, gn(after_err_label))
+        push!(stmts, err_label)
+        check_error = :($error("invoke: argument type error"))
```

Comment:
`TopNode` tells the compiler to resolve the symbol in the `Base` module relative to the module that is being defined. this is important when bootstrapping so that references to the old `Base` module don't get copied to the new `Base` module

effectively, `TopNode(sym) == getfield(base_relative_to(current_module()), sym)`